crates/adapters/src/catalog.rs

```diff
@@ -23,39 +145,134 @@ impl Catalog {
         Self::default()
     }
 
-    pub fn register_input_zset_handle<K, R>(&mut self, name: &str, handle: CollectionHandle<K, R>)
-    where
-        K: DBData + for<'de> Deserialize<'de>,
-        R: DBWeight + ZRingValue,
+    /// Add an input stream of Z-sets to the catalog.
+    pub fn register_input_zset<Z>(
+        &mut self,
+        name: &str,
+        stream: Stream<RootCircuit, Z>,
+        handle: CollectionHandle<Z::Key, Z::R>,
+    ) where
+        Z: ZSet + Send + Sync,
+        Z::R: ZRingValue + Into<i64> + Sync,
+        Z::Key: for<'de> Deserialize<'de> + Serialize + Sync,
     {
         self.register_input_collection_handle(name, DeZSetHandle::new(handle));
+
+        // Inputs are also outputs.
+        self.register_output_zset(name, stream);
     }
 
     /// Add a named input stream handle to the catalog.
-    pub fn register_input_collection_handle<H>(&mut self, name: &str, handle: H)
+    fn register_input_collection_handle<H>(&mut self, name: &str, handle: H)
     where
         H: DeCollectionHandle + 'static,
     {
         self.input_collection_handles
             .insert(name.to_owned(), Box::new(handle));
     }
 
-    /// Add a named output stream handle to the catalog.
-    pub fn register_output_batch_handle<H>(&mut self, name: &str, handle: H)
+    /// Add an output stream of Z-sets to the catalog.
```

Comment:
this should also be controlled by some configuration somewhere
the compiler could also generate this

Reply:
Yes, this is the main thing missing in this PR.  It's easy to add control knobs, but it's not clear who/how will set them. Some considerations:

- We need to start distinguishing between tables and streams in SQL (there is an issue for this). For streams, we probably should disable the two stateful queries (neighborhoods and quantiles), since streams don't have state.

- Even if a relation is declared as a table, but all queries over this table are stateless or use bounded time windows, we probably want to truncate state not used in any queries to avoid unbounded state.

- Finally, even for tables whose integral must be maintained indefinitely, these additional debugging outputs require memory + CPU to maintain, so we probably want to give users explicit control knobs to selectively enable/disable them. This can probably be done via some SQL annotations or dynamically, as part of the pipeline configuration.